File: cli/drand_draw/encode.py
CHAINS = {
    'quicknet': {
        'prefix': 'q',
        'genesis': 1692803367,
        'period': 3,
        'hash': '52db9ba70e0cc0f6eaf7803dd07447a1f5477735fd3f661792ba94600c84e971',
    },
    'default': {
        'prefix': 'd',
        'genesis': 1595431050,
        'period': 30,
        'hash': '8990e7a9aaed2ffed73dbd7092123d6f289930540d7651336225dc172e51b2ce',
    },
    'evmnet': {
        'prefix': 'e',
        'genesis': 1727521075,
        'period': 3,
        'hash': '04f1e9062b8a81f848fded9c12306733282b2727ecced50032187751166ec8c3',
    },
}

PREFIX_TO_CHAIN = {v['prefix']: k for k, v in CHAINS.items()}

_BASE36_CHARS = '0123456789abcdefghijklmnopqrstuvwxyz'


def to_base36(n):
    if n == 0:
        return '0'
    result = []
    while n > 0:
        result.append(_BASE36_CHARS[n % 36])
        n //= 36
    return ''.join(reversed(result))


def from_base36(s):
    return int(s, 36)


def encode_short_code(chain_name, deadline, n, prizes=None, winners=None):
    prefix = CHAINS[chain_name]['prefix']
    deadline_hex = format(deadline, 'x')
    n_b36 = to_base36(n)
    parts = [prefix, deadline_hex, n_b36]
    if prizes:
        parts.append(','.join(to_base36(p) for p in prizes))
        if winners:
            parts.append(','.join(to_base36(w) for w in winners))
    return '-'.join(parts)


def decode_short_code(code):
    parts = code.split('-')
    if len(parts) < 3:
        raise ValueError('Invalid short code: need at least 3 parts')

    prefix = parts[0]
    if prefix not in PREFIX_TO_CHAIN:
        raise ValueError(f'Unknown chain prefix: {prefix}')

    chain_name = PREFIX_TO_CHAIN[prefix]
    deadline = int(parts[1], 16)
    n = from_base36(parts[2])
    prizes = None
    winners = None

    if len(parts) >= 4 and parts[3]:
        prizes = [from_base36(x) for x in parts[3].split(',')]
    if len(parts) >= 5 and parts[4]:
        winners = [from_base36(x) for x in parts[4].split(',')]

    return {
        'chain': chain_name,
        'deadline': deadline,
        'n': n,
        'prizes': prizes or [],
        'winners': winners or [],
    }


import re

def _try_shortcode(val):
    """Direct short code."""
    try:
        return decode_short_code(val)
    except (ValueError, IndexError):
        return None


def _find_shortcode_in_text(val):
    """Find short code pattern in text."""
    m = re.search(r'\b[qde]-[0-9a-f]+-[0-9a-z]+(?:-[0-9a-z,]+)*\b', val)
    if m:
        try:
            return decode_short_code(m.group(0))
        except (ValueError, IndexError):
            pass
    return None


def _extract_url(val):
    """Extract URL and parse its hash/search params."""
    u = re.search(r'https?://[^\s]+', val)
    if u:
        try:
            from urllib.parse import urlparse, parse_qs
            parsed = urlparse(u.group(0))
            frag = parsed.fragment.lstrip('#')
            if frag.startswith('/?'):
                frag = frag[2:]
            if frag:
                qs = parse_qs(frag)
            else:
                qs = parse_qs(parsed.query)
            if 'chain' in qs and 'deadline' in qs and 'n' in qs:
                return {
                    'chain': qs['chain'][0],
                    'deadline': int(qs['deadline'][0]),
                    'n': int(qs['n'][0]),
                    'prizes': [int(x) for x in qs.get('prizes', [''])[0].split(',') if x] if 'prizes' in qs else [],
                    'winners': [int(x) for x in qs.get('winners', [''])[0].split(',') if x] if 'winners' in qs else [],
                }
        except Exception:
            pass
    return None


def _try_fragment(val):
    """Try to parse as hash/query fragment."""
    idx = max(val.rfind('#'), val.find('?'))
    if idx >= 0:
        frag = val[idx:]
        if frag.startswith('/?'):
            frag = frag[1:]
        if frag.startswith('?') or '=' in frag:
            from urllib.parse import parse_qs
            qs = parse_qs(frag.lstrip('?'))
            if 'chain' in qs and 'deadline' in qs and 'n' in qs:
                return {
                    'chain': qs['chain'][0],
                    'deadline': int(qs['deadline'][0]),
                    'n': int(qs['n'][0]),
                    'prizes': [int(x) for x in qs.get('prizes', [''])[0].split(',') if x] if 'prizes' in qs else [],
                    'winners': [int(x) for x in qs.get('winners', [''])[0].split(',') if x] if 'winners' in qs else [],
                }
    return None


def _try_raw(val):
    """Try raw key=value."""
    if '=' in val:
        from urllib.parse import parse_qs
        qs = parse_qs(val.lstrip('?'))
        if 'chain' in qs and 'deadline' in qs and 'n' in qs:
            return {
                'chain': qs['chain'][0],
                'deadline': int(qs['deadline'][0]),
                'n': int(qs['n'][0]),
                'prizes': [int(x) for x in qs.get('prizes', [''])[0].split(',') if x] if 'prizes' in qs else [],
                'winners': [int(x) for x in qs.get('winners', [''])[0].split(',') if x] if 'winners' in qs else [],
            }
    return None


def smart_parse(val):
    """Smart parse input text to extract draw parameters.

    Tries in order:
    1. Direct short code
    2. Short code in text
    3. URL extraction
    4. Hash/query fragment
    5. Raw key=value
    """
    if not val or not val.strip():
        return None
    val = val.strip()
    steps = [_try_shortcode, _find_shortcode_in_text, _extract_url, _try_fragment, _try_raw]
    for step in steps:
        result = step(val)
        if result is not None:
            return result
    return None

File: src/encode.js
import { CHAIN_SHORT, SHORT_CHAIN } from './chains.js'

export function encodeShortCode(params) {
  const parts = [SHORT_CHAIN[params.chain] || 'q']
  parts.push(params.deadline.toString(16))
  parts.push(params.n.toString(36))
  if (params.prizes && params.prizes.length) {
    parts.push(params.prizes.map(p => p.toString(36)).join(','))
  }
  if (params.winners && params.winners.length) {
    parts.push(params.winners.map(w => w.toString(36)).join(','))
  }
  return parts.join('-')
}

export function decodeShortCode(code) {
  const parts = code.split('-')
  if (parts.length < 3) return null
  const chainId = CHAIN_SHORT[parts[0]] || 'quicknet'
  const deadline = parseInt(parts[1], 16)
  if (isNaN(deadline)) return null
  const n = parseInt(parts[2], 36)
  if (isNaN(n)) return null
  const result = { chain: chainId, deadline, n, prizes: [], winners: [] }
  if (parts[3]) {
    result.prizes = parts[3].split(',').map(x => parseInt(x, 36)).filter(x => !isNaN(x))
  }
  if (parts[4]) {
    result.winners = parts[4].split(',').map(x => parseInt(x, 36)).filter(x => !isNaN(x))
  }
  return result
}

export function paramsToShortCode(params) {
  return encodeShortCode(params)
}

export function shortCodeToParams(code) {
  return decodeShortCode(code)
}

export function hashToParams(hash) {
  if (!hash) return null
  let h = hash.replace(/^#\/?\/?/, '')
  if (!h) return null
  h = h.replace(/^\/\?/, '')
  if (h.startsWith('?') || h.includes('=')) {
    const search = h.startsWith('?') ? h : '?' + h
    const usp = new URLSearchParams(search)
    const params = {}
    for (const [k, v] of usp) params[k] = v
    if (params.chain && params.deadline && params.n) {
      const decoded = {
        chain: params.chain,
        deadline: parseInt(params.deadline),
        n: parseInt(params.n),
        prizes: params.prizes ? params.prizes.split(',').map(x => parseInt(x)).filter(x => !isNaN(x)) : [],
        winners: params.winners ? params.winners.split(',').map(x => parseInt(x)).filter(x => !isNaN(x)) : [],
      }
      if (!isNaN(decoded.deadline) && !isNaN(decoded.n)) return decoded
    }
    return null
  }
  if (h.startsWith('verify/')) {
    const code = h.slice(7)
    return decodeShortCode(code)
  }
  const decoded = decodeShortCode(h)
  if (decoded) return decoded
  return null
}

export function smartParse(val) {
  if (!val) return null
  // 1. Direct short code (must look like one)
  if (/^[qde]-[0-9a-f]+-/.test(val)) {
    const d = decodeShortCode(val)
    if (d) return d
  }
  // 2. Find short code in text
  const m = val.match(/\b[qde]-[0-9a-f]+-[0-9a-z]+(?:-[0-9a-z,]+)*\b/)
  if (m) {
    const d = decodeShortCode(m[0])
    if (d) return d
  }
  // 3. Extract URL, parse its hash or search
  const u = val.match(/https?:\/\/[^\s]+/)
  if (u) {
    try {
      const url = new URL(u[0])
      const frag = url.hash.replace(/^#/, '')
      const p = hashToParams(frag || url.search.slice(1))
      if (p) return p
    } catch { /* invalid URL */ }
  }
  // 4. Try as hash/query fragment
  const qIdx = Math.max(val.lastIndexOf('#'), val.indexOf('?'))
  if (qIdx >= 0) {
    const p = hashToParams(val.slice(qIdx))
    if (p) return p
  }
  // 5. Raw key=value
  if (val.includes('=')) {
    const p = hashToParams(val)
    if (p) return p
  }
  return null
}

export function paramsToHash(params) {
  const usp = new URLSearchParams()
  usp.set('chain', params.chain)
  usp.set('deadline', String(params.deadline))
  usp.set('n', String(params.n))
  if (params.prizes && params.prizes.length) usp.set('prizes', params.prizes.join(','))
  if (params.winners && params.winners.length) usp.set('winners', params.winners.join(','))
  return '#/?' + usp.toString()
}
